### DISCUSSION TITLE
how to make all property notifiable without `ObservableProperty` attr

### DISCUSSION THREAD

**Discussion by heartacker** [CommunityToolkit/dotnet#374] [Q&A]
hi, all guys

how to make all property notifiable. I do not want to add ObservableProperty at all
```csharp
public AllProperitiesWantToBeNotifiableBydefaultClass: ObservableObject
{
    public Int Age{ get; set;}  // I do not want to add [ObservableProperty]      
}
```

thanks

**Comment by heartacker** [comment]
Is there any api to deal with?
thanks

**Comment by Sergio0694** [comment]
You can use `SetProperty`, as [the docs explain as well](https://docs.microsoft.com/dotnet/communitytoolkit/mvvm/observableobject) 🙂

**Reply by heartacker** [reply]
thanks, I notice that.

and in fact, what I want is as below in some way

```
[AllPropertiesObservableBydefaut]
public AllProperitiesWantToBeNotifiableBydefaultClass: ObservableObject
{
    public Int Age{ get; set;}  // I do not want to add [ObservableProperty]   
    ...
    ...   
    ...
	so many props
    public Int Name{ get; set;}
    
    [NotObservableProperty]
	public Int IdoWantThis{ get; set;}

    public Int fieldFoo;
	
}
```
If we can type the the simple property like `public Int Name{ get; set;}` ,We could change most code to wpf code

All we need to do is add `[AllPropertiesObservableBydefaut]` and `ObservableObject`

**Reply by heartacker** [reply]
or
```
[ObservableObject]
[AllPropertiesObservableBydefaut]
public AllProperitiesWantToBeNotifiableBydefaultClass
{
    public Int Age{ get; set;}  // I do not want to add [ObservableProperty]   
    ...
    ...   
    ...
	so many props
    public Int Name{ get; set;}
    
    [NotObservableProperty]
	public Int IdoWantThis{ get; set;}

    public Int fieldFoo;
	
}
```

will be better , 
and we could add some swithflag to csproj or some configfile in order to enable/disable codegen for some purpose

**Reply by Sergio0694** [reply]
What you're proposing is just not possible at all in C#, even if we wanted to. Generators cannot modify existing code 🙂

**Reply by heartacker** [reply]
can this be possible? That is enough at least
```
[AllPropertiesObservableBydefaut]
public AllProperitiesWantToBeNotifiableBydefaultClass: ObservableObject
{
    public Int Age{ get; set;}  // I do not want to add [ObservableProperty]   
    ...
    ...   
    ...
	so many props
    public Int Name{ get; set;}
    
    [NotObservableProperty]
	public Int IdoWantThis{ get; set;}

    public Int fieldFoo;
	
}
```

**Reply by Sergio0694** [reply]
No, this is not possible at all. You have a property there, and you'd like the generator to _change_ it and make it be observable. Generators cannot do this at all: they can only _add_ code. We do want to move to properties in the future, which would get somewhat closer to this, but (1) we need support for that, so not sooner than C# 12 at the earliest, and (2) we'd still keep `[ObservableProperty]` mandatory over each property. Adding it over a class would make code harder to read, as it would no longer be immediately clear which properties are observable and which ones are not.

**Reply by heartacker** [reply]
> No, this is not possible at all. You have a property there, and you'd like the generator to _change_ it and make it be observable. Generators cannot do this at all: they can only _add_ code. We do want to move to properties in the future, which would get somewhat closer to this, but (1) we need support for that, so not sooner than C# 12 at the earliest, and (2) we'd still keep `[ObservableProperty]` mandatory over each property. Adding it over a class would make code harder to read, as it would no longer be immediately clear which properties are observable and which ones are not.

thanks

and I found this. seem usful https://github.com/Fody/PropertyChanged#overview

**Reply by heartacker** [reply]
Fody.PropertyChanged  Works well.
Could you please tell me the difference between the mvvmtoolskits and fody property changed?

Fody.propertychanged can generate the code based on property  But according to your description, We are unable to do that

And can do this like fody 's way？thanks
